### DISCUSSION TITLE
how to bind to a network card

### DISCUSSION THREAD

**Discussion by lightyear67** [cesanta/mongoose#2298] [Ask questions, get answers] [answered]
My PC has multiple network cards, how to select one of them when I establishing TCP or UDP connections. 
This may be a simple question, but it's bothering me now

**Comment by scaprile** [comment, marked as answer]
I suggest you read some networking.
Your PC has an operating system
Your OS has a TCP/IP stack
Your cards are connected to networks
Each network has a network address and each card has a host address
When you start a listener, you choose on what address you will listen. This you can control. See the examples.
When you start a connection, your stack decides on which interface to exit based on the destination address and its routing table. This you can't, and usually don't need to.
In IPv6 there are other options we do not support

**Reply by lightyear67** [reply]
Thank you very much for your answer. I‘m not good at this.
I want to send data to a multicast address, which works fine, but how can I bind to the specified network card. Or should I not do this?
This is my code.
`
static const char *addr = "udp://239.255.255.250:19000";

int main(void) {

	struct mg_mgr mgr;

	static struct mg_connection *c;

	mg_mgr_init(&mgr);

	c = mg_connect(&mgr, addr, fn, NULL);

	mg_timer_add(&mgr, 2000, MG_TIMER_REPEAT | MG_TIMER_RUN_NOW, tfn, c);

	while (true) mg_mgr_poll(&mgr, 200);

	mg_mgr_free(&mgr);

	return 0;
}
`

**Reply by scaprile** [reply]
Multicast is a totally different beast and depends entirely on your OS, that was not your question, You don't bind to a card, you send to a multicast address (Please check working examples and read their tutorials
https://mongoose.ws/documentation/tutorials/udp-ssdp-search/ ) or listen on a multicast address, which, as I said above, requires OS support

I'm not sure right now, I think I already answered something like this for someone wanting to do an Alpaca device... search for that (either questions or issues, can't remember)

This is what I do in Linux, I don't do Win:
```
static const char *s_ssdp_url = "udp://0.0.0.0:1900";
...
  mg_mgr_init(&mgr);
  c = mg_listen(&mgr, s_ssdp_url, fn, NULL);
//
    struct ip_mreq mreq;
    mreq.imr_multiaddr.s_addr = inet_addr("239.255.255.250");
    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
    if (setsockopt( (int) c->fd, IPPROTO_IP, IP_ADD_MEMBERSHIP, (char*) &mreq, sizeof(mreq)) < 0)
        return 1;
//
```
This basically listens on all interfaces at UDP port 1900, then, you ask your OS to allow binding the proper socket to a multicast group 
To use a specific card/interface you should use that ifc IP instead, but I don't know if that would work, do your homework, try this first. Good learning.

**Reply by lightyear67** [reply]
It has helped me a lot, I understand it now. Thank you for your timely reply, and I have learned a lot.
